### DISCUSSION TITLE
read modules endpoints

### DISCUSSION THREAD

**Discussion by tomersolomon1** [fybrik/fybrik#266] [Developers]
The purpose of this discussion is to discuss the feature called "read modules endpoints": requirements, initial proposals, technology, etc.

The basic idea and motivation is that modules need to communicate with one another. The focus of this feature is help in the case of reading data from other module. Say module A wants to read data from module B. If module A “knows” how to access module B, and has the connectivity permissions to do so, then there is no problem.
This is the case for example if the address of B is pre-defined or configured in set-up time. For example, the secret-provider has a such address, secret-provider.m3d-system.svc.cluster.local, which is defined in the manager’s configmap.

However, this is not the case for example for the arrow flight module. This module is created only in runtime, using address which is not predefined in advance. Instead, at least currently, a random namespace is created, and the pods of the arrow flight module are deployed over there. Moreover, currently the user needs to figure out the address of the module in order communicate with that module after it was deployed. Figuring out where the module is deployed is currently done manually be inspecting the status of the m4d-application. Obviously, this is less desired.
Instead, we probably want to use some sort of virtualization / redirection. In this approach, we create a virtualized address for the module. This address can be fixed / configured beforehand somehow. So, when module A wants to read data from module B, it reaches to the “fixed” virtualized endpoint, instead of talking to service B directly.

Some notes and questions:
1.	The design should also support multi-cluster settings. Ideally, the design for single cluster and multi-cluster should have same spirit / ingredients.
2.	Can we use Istio or rely only on Kubernetes native capabilities?
3.	Where is the redirection to the “real” address of module B happens? For example, if we use Istio, this redirection can be done either in the sidecar or in the egress gateway.
4.	How to augment the m4d-application to support this feature? Maybe add to it the virtualized address the workload is expecting to access? For example, the arrow-flight module..

This blog: https://cloud.google.com/blog/products/gcp/kubernetes-best-practices-mapping-external-services explains several ideas in case module B is in a different cluster. For example, we can use k8s `ExternalName` service in case the module is in a different cluster.
https://stackoverflow.com/questions/37221483/service-located-in-another-namespace claims we can use `ExternalName` also in case the url we're trying to reach is in the same cluster.
Alternatively, if Istio is on the table, we can create a `VirtualService` and that’s it?

#### Edit
I checked `ExternalName` and it can point also to services in the same cluster. For example, created the following service:
```
kind: Service
apiVersion: v1
metadata:
 name: secret-provider-external-name-service
spec:
 type: ExternalName
 externalName: secret-provider.m4d-system.svc.cluster.local
```

And after deploying it I was able to access the secret-provider by
```
curl 'http://secret-provider-external-name-service.default:5555/get-secret/....'
```

**Comment by hunchback** [comment]
I think we should not require istio, and should provide a solution independent of istio

**Reply by shlomitk1** [reply]
We have an issue regarding deploying blueprints in a single "pre-defined" namespace, so finding an endpoint of the flight module won't be  an issue. It might help, though, for the multiple cluster environment in case the endpoint location is unknown. It's unlikely that the read module will not be deployed next to the workload but still may happen.

**Comment by roee88** [comment]
@tomersolomon1 let's start with the endpoint for a read module to be used by application workloads. Even after moving all modules to a single namespace, the code that the user workload runs is not aware of the endpoint to use. Currently the manager:
1. Lets the module decide on endpoint (service name). With a move to single namespace we may also require it to consider the Helm release name that the manager chooses
2. Provide details in rendered helm NOTES.txt and reflect that in the M4DApplicaiton status 

There are several issues here, especially when we consider workloads that are not just notebooks but rather compiled apps. The main things that I thought about are:
1. How do we allow applications to be written without a need to read the M4DApplicaiton status. This could be by providing a well defined endpoint (single or one that considers the asset) that the application code can construct a priori. Or it could be by providing the workload with a configmap with the endpoint information, or any other method that we can come up with.
1. Know the endpoint before actually installing the module. This will also allow us to align with the multicluster deployment model of tools other than Razee (and to align with ongoing [efforts](https://docs.google.com/document/d/1cWcdB40pGg3KS1eSyb9Q6SIRvWVI8dEjFp9RI0Gk0vg/edit#heading=h.u7jfy9wqpd2b) in sig-multicluster) where you don't actually read the plain status of remotely applied resources (the status you do get reflects the reconcile of the resource instead)

**Reply by shlomitk1** [reply]
While I agree with the endpoint construction, how can the data user know when the data is ready without reading the status?

**Reply by roee88** [reply]
I agree that it's more complex, but readiness is easier than custom status. Long term we can even add a sidecar container than checks for readiness or even try using a readiness probe directly on the application container.

**Comment by roee88** [comment]
Re external name services, if we can get a working solution with that and without Istio it would be great. I'll later share an old attempt we had using Istio. Regardless, please consider limitations to the service name (e.g. length) and define how an endpoint would look like.

**Comment by tomersolomon1** [comment]
@roee88 @shlomitk1 Could you please clarify what you mean by a "single namespace"? How is the namespace defined, say as a function of the m4d-application name / blueprint name? Could it be somehow pre-defined, or it's random by-design? What component decides about the name of that namespace?

@roee88 I think both methods you suggested are good (get the svc name from configmap or make it pre-defined).
However, in my opinion, it would be easier for users if they could give us the name of the service they expect. And it would be even better if it would have a default, for example `arrow-flight-module`. Moreover, in my understanding, the biggest question is actually about the **namespace** where that `ExternalName` service would be (if we use `ExternalName` servcie for example), rather than the service name itself. We have a few options:
1. Use the same namespace as the m4d-application
2. Let the user decide (let it be a field in the m4d-application for example), so the user can decide on the name of the service + the namespace.

Regarding the use of Istio - I agree that if we don't plan to use Istio at all in the project, we shouldn't probably use it here. However, if we do plan to use Istio eventually, I think it might make our lives easier in the long run to use it here as well. For example, I'm afraid we might run to some problems with SNI or something. I'm not 100% sure. And besides, I suppose it's more elegant to use Istio to solve all our problems, and configure all connectivity / service virtuallization in one central place, instead of solving one issue here and one issue there.

**Reply by roee88** [reply]
See https://github.com/IBM/the-mesh-for-data/issues/257 as it's highly relevant here.

**Comment by tomersolomon1** [comment]
I had a discussion with @ronenkat about the feature, to summarize:
1. Regarding configuration - without loss of generality, each asset can have a different endpoint. We'll combine the approaches we discussed above - the user will be able to specify the name of the endpoint it expects for each asset (will provide a default). After it was deployed, a configmap will be created containing all information.
2. Regarding the redirection / virtualization itself - we discussed the possibility of making it plugable via a module. In other words, a "configuration" module will be added to the blueprint. This module will be in charged of configuring the virtualized endpoint. This way, in case we have Istio we can use Istio, and tomorrow we'll use Linkerd, and otherwise we use native k8s capabilities. This module will be configured with the details of the endpoints.
3. We not sure how to generalize it to the case of multi-cluster data-path. We should discuss it with @froesef

**Reply by roee88** [reply]
Starting a design document may fit now. 

Re 1 - this is one of the things I had in mind but rather than endpoint I imagined a name (used as a dns label in the endpoint). 
Re 2 - what if we can get the functionality without dependencies? (i.e. Kubernetes only) Would it still apply?

**Reply by ronenkat** [reply]
> Re 2 - what if we can get the functionality without dependencies? (i.e. Kubernetes only) Would it still apply?
I believe so as it would be a loosely coupled solution between the code in the manager and the implementation. I would envision that doing something across cluster may be implementation specific.

**Reply by roee88** [reply]
Got it, thanks.

**Comment by tomersolomon1** [comment]
See following links:
Configuring the workload to use the right URL:
https://docs.google.com/document/d/10qL_sUBqbSmPww1GQ76ON8KjQjFVByU3aXarmCx9uHA/edit?usp=sharing

Configuring the environment to allow such connectivity: 
https://docs.google.com/document/d/1KzkzzE4BnxnZb6PEiJOK3ol3bygt8N24z7V8CfV0Pd8/edit

**Comment by simanadler** [comment]
@tomersolomon1 @roee88 I think it would be really helpful to have a description of how we want this to work from the viewpoint of the users.  I think we have several different users:
1) Notebook user writing new code
2) Owner of an existing application that currently accesses data via jdbc, s3, ...
3) Developer of a new compiled application
4) M4D module developer that wants to evaluate or transform data, and wants to use the read module to obtain the data from storage into memory (ex: into a panda)

How would the solution look for each of these different users?
